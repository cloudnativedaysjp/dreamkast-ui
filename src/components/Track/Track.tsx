import React, { useEffect, useState, useCallback, useRef } from 'react'
import { IvsPlayer } from '../IvsPlayer'
import { Chat } from '../Chat'
import Grid from '@material-ui/core/Grid'
import {
  Track,
  Talk,
  TalkApi,
  Configuration,
  Profile,
  Event,
} from '../../client-axios'
import { TalkSelector } from '../TalkSelector'
import { TalkInfo } from '../TalkInfo'
import { Sponsors } from '../Sponsors'
import ActionCable from 'actioncable'
import dayjs from 'dayjs'
import 'dayjs/locale/ja'

type Props = {
  event: Event
  profile?: Profile
  selectedTrack?: Track
  propTalks?: Talk[]
}

export const TrackView: React.FC<Props> = ({
  event,
  profile,
  selectedTrack,
  propTalks,
}) => {
  const [talks, setTalks] = useState<Talk[]>(propTalks ? propTalks : [])
  const [videoId, setVideoId] = useState<string>()
  const [selectedTalk, setSelectedTalk] = useState<Talk>()
  const [timer, setTimer] = useState<number>()
  const [isLiveMode, setIsLiveMode] = useState<boolean>(true)
  const [chatCable, setChatCable] = useState<ActionCable.Cable | null>(null)
  const beforeTrackId = useRef<number | undefined>(selectedTrack?.id)

  const findDayId = () => {
    const today = dayjs(new Date()).tz('Asia/Tokyo').format('YYYY-MM-DD')
    let dayId = ''
    event?.conferenceDays?.forEach((day) => {
      if (day.date == today && day.id) {
        dayId = String(day.id)
      }
    })
    return dayId
  }

  const getTalks = useCallback(async () => {
    const api = new TalkApi(
      new Configuration({ basePath: window.location.origin }),
    )
    const dayId = findDayId()
    if (!dayId) return
    const { data } = await api.apiV1TalksGet(
      !!event ? event?.abbr : '',
      String(selectedTrack?.id),
      dayId,
    )
    setTalks(data)
  }, [event, selectedTrack])

  useEffect(() => {
    if (!propTalks) getTalks()
  }, [getTalks])

  useEffect(() => {
    if (isLiveMode) getTalks()
  }, [isLiveMode])

  const selectTalk = (talk: Talk) => {
    if (!talk.onAir) {
      setIsLiveMode(false)
    }
    setSelectedTalk(talk)
    setVideoId(talk.onAir ? selectedTrack?.videoId : talk.videoId)
  }

  useEffect(() => {
    if (
      !talks.length ||
      (!isLiveMode && beforeTrackId.current === selectedTrack?.id)
    )
      return
    beforeTrackId.current = selectedTrack?.id
    const onAirTalk = talks.find((talk) => talk.onAir)
    setSelectedTalk(onAirTalk ? onAirTalk : talks[0])
    setVideoId(onAirTalk ? selectedTrack?.videoId : talks[0].videoId)
  }, [talks])

  const actionCableUrl = () => {
    if (window.location.protocol == 'http:') {
      return `ws://${window.location.host}/cable`
    } else {
      return `wss://${window.location.host}/cable`
    }
  }

  const onChecked = (
    _event: React.ChangeEvent<HTMLInputElement>,
    checked: boolean,
  ) => {
    setIsLiveMode(checked)
  }

  useEffect(() => {
    if (!selectedTrack) return
    if (chatCable) chatCable.disconnect()
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const actionCable = require('actioncable')
    const wsUrl = actionCableUrl()
    const cable = actionCable.createConsumer(wsUrl)
    setChatCable(cable)
    cable.subscriptions.create(
      { channel: 'OnAirChannel', eventAbbr: event?.abbr },
      {
        received: (msg: { [trackId: number]: Talk }) => {
          getTalks() // onAirの切り替わった新しいTalk一覧を取得
          if (!msg[selectedTrack.id] || !selectedTalk) return
          if (
            selectedTrack.id == msg[selectedTrack.id].trackId &&
            selectedTalk.id != msg[selectedTrack.id].id
          ) {
            setSelectedTalk(msg[selectedTrack.id])
            setVideoId(msg[selectedTrack.id].videoId)
          }
        },
      },
    )
  }, [selectedTrack])

  useEffect(() => {
    clearInterval(timer)
    setTimer(
      window.setInterval(() => {
        window.tracker.track('watch_video', {
          track_name: selectedTrack?.name,
          talk_id: selectedTalk?.id,
          talk_name: selectedTalk?.title,
        })
      }, 120 * 1000),
    )
  }, [selectedTrack, selectedTalk])

  return (
    <Grid container spacing={0} justify="center" alignItems="flex-start">
      <Grid item xs={12} md={8}>
        <IvsPlayer playBackUrl={videoId} autoplay={true}></IvsPlayer>
        <Sponsors event={event} />
      </Grid>
      <Grid item xs={12} md={4}>
        <Chat event={event} profile={profile} talk={selectedTalk} />
      </Grid>
      <Grid item xs={12} md={8} style={{ height: '100%' }}>
        <TalkInfo
          event={event}
          selectedTalk={selectedTalk}
          selectedTrackName={selectedTrack?.name}
          selectedTrackId={selectedTrack?.id}
        />
      </Grid>
      <Grid item xs={12} md={4} style={{ height: '100%' }}>
        <TalkSelector
          selectedTalk={selectedTalk}
          selectedTrackId={selectedTrack?.id}
          talks={talks}
          isLiveMode={isLiveMode}
          changeLiveMode={onChecked}
          selectTalk={selectTalk}
        />
      </Grid>
    </Grid>
  )
}
