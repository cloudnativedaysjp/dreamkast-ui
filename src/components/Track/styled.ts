import styled from 'styled-components'
import { Button } from '@material-ui/core'

export const ShareButton = styled(Button)`
  margin-right: 5px;
  margin-bottom: 5px;
  width: 20%;
  height: 50px;
  border: none !important;
  background: rgba(255, 255, 255, 0.7);
  &:hover {
    background-color: #ffffff;
  }
`
