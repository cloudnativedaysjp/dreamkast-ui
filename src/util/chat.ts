import {
  ChatMessage as ChatMessageInterface,
  ChatMessageMessageTypeEnum,
} from '../client-axios'

export class ChatMessageMap extends Map<number, ChatMessageClass> {
  addMessage = (msg: ChatMessageClass): void => {
    if (!msg.id) return

    if (msg.replyTo) {
      const parent = this.get(msg.replyTo)
      if (parent) {
        if (!parent.children) {
          parent.children = new Map<number, ChatMessageClass>()
        }
        parent.children.set(msg.id, msg)
      } else {
        this.set(msg.id, msg)
      }
    } else {
      this.set(msg.id, msg)
    }
  }
}

export class ChatMessageClass implements ChatMessageInterface {
  id?: number
  profileId?: number
  speakerId?: number
  eventAbbr: string
  roomId?: number
  roomType?: string
  createdAt?: string
  body: string
  messageType: ChatMessageMessageTypeEnum
  replyTo?: number
  children?: Map<number, ChatMessageClass>

  constructor(
    id: number,
    profileId: number,
    speakerId: number,
    eventAbbr: string,
    roomId: number,
    roomType: string,
    createdAt: string,
    body: string,
    messageType: ChatMessageMessageTypeEnum,
    replyTo: number,
  ) {
    this.id = id
    this.profileId = profileId
    this.speakerId = speakerId
    this.eventAbbr = eventAbbr
    this.roomId = roomId
    this.roomType = roomType
    this.createdAt = createdAt
    this.body = body
    this.messageType = messageType
    this.replyTo = replyTo
    this.children = new Map<number, ChatMessageClass>()
  }
}
