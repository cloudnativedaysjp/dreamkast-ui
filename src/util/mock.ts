import { Event, Talk, Track } from '../client-axios'

export const MockEvent: Event = {
  id: 1,
  name: 'Observability Conference 2022',
  abbr: 'o11y2022',
  status: 'open',
  theme: 'this is theme',
  about: 'this is about',
  privacy_policy: 'this is privacy policy',
  privacy_policy_for_speaker: 'this is privacy policy for speaker',
  copyright: 'this is copyright',
  coc: 'this is coc',
}

export const Talks: Talk[] = [
  {
    id: 1,
    trackId: 1,
    videoId: '453942665',
    title: 'ほげふが',
    abstract: 'ぴよぴよ',
    speakers: [
      { id: 1, name: 'alice' },
      { id: 2, name: 'bob' },
    ],
    dayId: 1,
    showOnTimetable: true,
    startTime: '2021-02-15 14:00:00',
    endTime: '2021-02-15 15:00:00',
    talkDuration: 1,
    talkDifficulty: 'hoge',
    talkCategory: 'fuga',
  },
  {
    id: 2,
    trackId: 1,
    videoId: '450845161',
    title: 'ふー',
    abstract: 'ばー',
    speakers: [
      { id: 3, name: 'taro' },
      { id: 4, name: 'jiro' },
    ],
    dayId: 1,
    showOnTimetable: true,
    startTime: '2021-02-15 14:00:00',
    endTime: '2021-02-15 15:00:00',
    talkDuration: 1,
    talkDifficulty: 'hoge',
    talkCategory: 'fuga',
    onAir: true,
  },
  {
    id: 3,
    trackId: 1,
    videoId: '453963160',
    title: 'ほげほげほげ',
    abstract: 'ぴよぴよぴよ',
    speakers: [
      { id: 5, name: 'foo' },
      { id: 6, name: 'bar' },
    ],
    dayId: 1,
    showOnTimetable: true,
    startTime: '2021-02-15 14:00:00',
    endTime: '2021-02-15 15:00:00',
    talkDuration: 1,
    talkDifficulty: 'hoge',
    talkCategory: 'fuga',
  },
]

export const Tracks: Track[] = [
  {
    id: 10,
    name: 'A',
    videoId: '453942665',
    videoPlatform: 'vimeo',
  },
  {
    id: 11,
    name: 'B',
    videoId: '450845161',
    videoPlatform: 'vimeo',
  },
  {
    id: 12,
    name: 'C',
    videoId: '453963160',
    videoPlatform: 'vimeo',
  },
]
