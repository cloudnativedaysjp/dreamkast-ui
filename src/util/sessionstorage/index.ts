// https://developer.mozilla.org/ja/docs/Web/API/Web_Storage_API/Using_the_Web_Storage_API

export function isStorageAvailable(storageType: string): boolean | undefined {
  let storage
  try {
    storage = window[storageType]
    const x = '__storage_test__'
    storage.setItem(x, x)
    storage.removeItem(x)
    return true
  } catch (e) {
    return (
      e instanceof DOMException &&
      // everything except Firefox
      (e.code === 22 ||
        // Firefox
        e.code === 1014 ||
        // test name field too, because code might not be present
        // everything except Firefox
        e.name === 'QuotaExceededError' ||
        // Firefox
        e.name === 'NS_ERROR_DOM_QUOTA_REACHED') &&
      // acknowledge QuotaExceededError only if there's something already stored
      storage &&
      storage.length !== 0
    )
  }
}
