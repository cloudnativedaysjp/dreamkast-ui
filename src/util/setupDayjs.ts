import dayjs from 'dayjs'
import utc from 'dayjs/plugin/utc'
import timezone from 'dayjs/plugin/timezone'

export const setupDayjs = () => {
  dayjs.extend(timezone)
  dayjs.extend(utc)
  dayjs.tz.setDefault('Asia/Tokyo')
}
