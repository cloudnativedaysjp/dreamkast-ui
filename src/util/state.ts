import { useState } from 'react'

//export const useStateWithLocalStorage = <T>(key: string): ([T, (value: T | ((val: T) => T)) => void]) => {
export default function useStateWithLocalStorage<T>(
  key: string,
): [T, (value: T | ((val: T) => T)) => void] {
  const [storedValue, setStoredValue] = useState<T>(() => {
    const item = window.localStorage.getItem(key)
    //return item ? JSON.parse(item) : initialValue;
    return item ? JSON.parse(item) : ''
  })

  const setValue = (value: T | ((val: T) => T)) => {
    const valueToStore = value instanceof Function ? value(storedValue) : value
    setStoredValue(valueToStore)
    window.localStorage.setItem(key, JSON.stringify(valueToStore))
  }

  return [storedValue, setValue]
}
