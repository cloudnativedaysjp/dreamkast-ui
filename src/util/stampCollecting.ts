import { Talk } from '../generated/dreamkast-api.generated'
import { createHash } from 'crypto'
import { isStorageAvailable } from './sessionstorage'

export function getSlotId(talk: Talk): number {
  if (talk.dayId == null || talk.slotNum == null) {
    return 0
  }
  return talk.dayId * 100 + talk.slotNum
}

export function getPointEventIdBySlot(salt: string, slotId: number): string {
  return getPointEventId(salt, 100000 + slotId)
}

export function getPointEventId(salt: string, eventNum: number): string {
  const shasum = createHash('sha1')
  return shasum.update(`${salt}/${eventNum}`).digest('hex')
}

export function makeTrackResolveMap(eventAbbr: string) {
  const trackId = [0, 1, 2, 3, 4, 5]
  return trackId.reduce((accum, curr) => {
    const eventNum = 130001 + curr
    accum[getPointEventId(eventAbbr, eventNum)] = curr
    return accum
  }, {} as Record<string, number>)
}

export type QRCodeRequestResult = 'ok' | 'skipped' | 'invalid' | 'error'
const keyQRCodeStampResult = 'qrCodeStampResult'
const keyTrailMapOpenNext = 'trailMapOpenNext'

export const setQRCodeStampResult = (res: QRCodeRequestResult) => {
  if (isStorageAvailable('sessionStorage')) {
    sessionStorage.setItem(keyQRCodeStampResult, res)
  }
}

export const clearQRCodeStampResult = () => {
  if (isStorageAvailable('sessionStorage')) {
    sessionStorage.removeItem(keyQRCodeStampResult)
  }
}

export const getQRCodeStampResult = (): QRCodeRequestResult | null => {
  if (isStorageAvailable('sessionStorage')) {
    const res = sessionStorage.getItem(keyQRCodeStampResult)
    if (res === 'ok' || res === 'skipped' || res === 'invalid' || res == null) {
      return res
    }
  }
  return 'error'
}

export const setTrailMapOpenNext = () => {
  if (isStorageAvailable('sessionStorage')) {
    sessionStorage.setItem(keyTrailMapOpenNext, 'y')
  }
}

export const clearTrailMapOpenNext = () => {
  if (isStorageAvailable('sessionStorage')) {
    sessionStorage.removeItem(keyTrailMapOpenNext)
  }
}

export const getTrailMapOpenNext = (): boolean => {
  if (isStorageAvailable('sessionStorage')) {
    return !!sessionStorage.getItem(keyTrailMapOpenNext)
  }
  return false
}

