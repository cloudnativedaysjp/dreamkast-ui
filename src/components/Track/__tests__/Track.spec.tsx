import React from 'react'
import renderer from 'react-test-renderer'
import { Talks } from '../../../util/mock'
import { TrackView } from '../Track'

test('Track', () => {
  const component = renderer.create(
    <TrackView propTalks={Talks} selectedTrackId={1} />,
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
