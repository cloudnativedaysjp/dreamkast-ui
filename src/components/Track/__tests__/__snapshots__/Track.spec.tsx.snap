// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Track 1`] = `
<div
  className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-align-items-xs-flex-start MuiGrid-justify-xs-center"
>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-8"
  >
    <iframe
      allow="autoplay; fullscreen"
      allowFullScreen={true}
      className="makeStyles-player-1"
      frameBorder="0"
      src="https://player.vimeo.com/video/undefined?autoplay=0"
    />
  </div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-3"
  >
    <div>
      <h2>
        Chat / QA
      </h2>
      <h3 />
      <ul
        className="MuiList-root MuiList-padding"
      >
        <div
          className="MuiListItemText-root"
        >
          <span
            className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            foooo
          </span>
        </div>
        <div
          className="MuiListItemText-root"
        >
          <span
            className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            bar
          </span>
        </div>
        <div
          className="MuiListItemText-root"
        >
          <span
            className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            baz
          </span>
        </div>
        <div
          className="MuiListItemText-root"
        >
          <span
            className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            hoge
          </span>
        </div>
        <div
          className="MuiListItemText-root"
        >
          <span
            className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            fuga
          </span>
        </div>
        <div
          className="MuiListItemText-root"
        >
          <span
            className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            piyo
          </span>
        </div>
        <div
          className="MuiListItemText-root"
        >
          <span
            className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            fooo
          </span>
        </div>
      </ul>
    </div>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-10 MuiGrid-grid-md-8"
  >
    <section
      className="sc-bdfBwQ"
    >
      <h2 />
      <h3 />
      <div />
    </section>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-3"
  >
    <div
      className="sc-gsTCUz lmYqBg"
    >
      <div
        className="sc-dlfnbm iRYuVZ"
      >
        このトラックのセッション
      </div>
      <ul
        className="MuiList-root sc-hKgILt fWMRiy MuiList-padding"
      >
        <div
          aria-disabled={false}
          className="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex={0}
        >
          <div
            className="MuiListItemText-root MuiListItemText-inset"
          >
            <span
              className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
            >
              ほげふが
            </span>
          </div>
        </div>
        <div
          aria-disabled={false}
          className="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex={0}
        >
          <div
            className="MuiListItemText-root MuiListItemText-inset"
          >
            <span
              className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
            >
              ふー
            </span>
          </div>
        </div>
        <div
          aria-disabled={false}
          className="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex={0}
        >
          <div
            className="MuiListItemText-root MuiListItemText-inset"
          >
            <span
              className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
            >
              ほげほげほげ
            </span>
          </div>
        </div>
      </ul>
    </div>
  </div>
</div>
`;
